flocker/node/agents/test/test_ebs.py

```diff
@@ -31,6 +53,133 @@ def test_repr(self):
             repr(AttachedUnexpectedDevice(requested, discovered))
         )
 
+    def test_nothing_discovered_repr(self):
+        """
+        If no device is discovered, the repr of ``AttachedUnexpectedDevice``
+        shows this with a ``None`` value for ``discovered``.
+        """
+        requested = FilePath(b"/dev/sda")
+        discovered = None
+        expected = (
+            "AttachedUnexpectedDevice(requested='/dev/sda', discovered=None)"
+        )
+        self.assertEqual(
+            expected,
+            repr(AttachedUnexpectedDevice(requested, discovered))
+        )
+
+    def test_wrong_requested_type(self):
+        """
+        If ``AttachedUnexpectedDevice.__init__`` is given something other than
+        an instance of ``FilePath`` for the value of ``requested``,
+        ``TypeError`` is raised.
+        """
+        self.assertRaises(
+            TypeError, AttachedUnexpectedDevice, object(), FilePath(b"/")
+        )
+
+    def test_wrong_discovered_type(self):
+        """
+        If ``AttachedUnexpectedDevice.__init__`` is given something other than
+        ``None`` or an instance of ``FilePath`` for the value of ``requested``,
+        ``TypeError`` is raised.
+        """
+        self.assertRaises(
+            TypeError, AttachedUnexpectedDevice, FilePath(b"/"), object(),
+        )
+
+
+class AttachVolumeAndWaitTests(SynchronousTestCase):
+    """
+    Tests for ``_attach_volume_and_wait_for_device``.
+    """
+    def setUp(self):
+        self.volume = BlockDeviceVolume(
+            dataset_id=uuid4(),
+            blockdevice_id=u"opaque storage backend id",
+            size=int(GiB(100).to_Byte()),
+        )
+        self.compute_id = u"opaque compute backend id"
+
+    @given(device=device_path, attach_attempt=integers(min_value=1))
+    def test_unexpected_attach_exception(self, device, attach_attempt):
+        """
+        If the ``attach_volume`` function raises an unexpected exception, it is
+        passed through.
+        """
+        def attach_volume(blockdevice_id, compute_id, device):
+            raise CustomException("Fake failure generated by test")
+
+        def detach_volume(*a, **kw):
+            pass
+
+        self.assertRaises(
+            CustomException,
+            _attach_volume_and_wait_for_device,
+            self.volume, self.compute_id, attach_volume, detach_volume,
+            device,
```

Comment:
slight preference for using kwargs here rather than positional (as you do for blockdevices) I had to go back to the definition a lot to determine what each of these arguments were.